Umbraco.ModelsBuilder/Configuration/Config.cs

```diff
@@ -165,6 +170,27 @@ private Config()
                 ModelsDirectory = GetModelsDirectory(root, value, AcceptUnsafeModelsDirectory);
             }
 
+            // default: initialized above with DefaultBinDirectory const
+            value = ConfigurationManager.AppSettings[prefix + "BinDirectory"];
+            if (!string.IsNullOrWhiteSpace(value))
+            {
+                var root = HostingEnvironment.IsHosted
+                    ? HostingEnvironment.MapPath("~/")
+                    : Directory.GetCurrentDirectory();
+                if (root == null)
+                    throw new ConfigurationErrorsException("Could not determine root directory.");
+
+                var path = HostingEnvironment.IsHosted
+                    ? HostingEnvironment.MapPath(value)
+                    : Path.GetFullPath(value);
+
+                if (path != null && path.StartsWith(root))
```

Comment:
@zpqrtbnk this part checks if the path is within the root. Though additional rights may be needed if the application's rights are restricted.